arch/arm64/core/smp.c

```diff
@@ -72,40 +73,39 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	master_core_mpid = MPIDR_TO_CORE(GET_MPIDR());
 
 	cpu_count = ARRAY_SIZE(cpu_node_list);
-	__ASSERT(cpu_count == CONFIG_MP_MAX_NUM_CPUS,
-		"The count of CPU Cores nodes in dts is not equal to CONFIG_MP_MAX_NUM_CPUS\n");
-
-	for (i = 0, j = 0; i < cpu_count; i++) {
-		if (cpu_node_list[i] == master_core_mpid) {
-			continue;
-		}
-		if (j == cpu_num - 1) {
-			cpu_mpid = cpu_node_list[i];
-			break;
-		}
-		j++;
-	}
-	if (i == cpu_count) {
-		printk("Can't find CPU Core %d from dts and failed to boot it\n", cpu_num);
-		return;
-	}
+	__ASSERT(cpu_count >= CONFIG_MP_MAX_NUM_CPUS,
+		"The count of CPU Core nodes in dts is not greater or equal to CONFIG_MP_MAX_NUM_CPUS\n");
 
 	arm64_cpu_boot_params.sp = Z_KERNEL_STACK_BUFFER(stack) + sz;
 	arm64_cpu_boot_params.fn = fn;
 	arm64_cpu_boot_params.arg = arg;
 	arm64_cpu_boot_params.cpu_num = cpu_num;
 
-	barrier_dsync_fence_full();
+	for (; i < cpu_count; i++) {
+		if (cpu_node_list[i] == master_core_mpid) {
+			continue;
+		}
+
+		cpu_mpid = cpu_node_list[i];
+
+		barrier_dsync_fence_full();
 
-	/* store mpid last as this is our synchronization point */
-	arm64_cpu_boot_params.mpid = cpu_mpid;
+		/* store mpid last as this is our synchronization point */
+		arm64_cpu_boot_params.mpid = cpu_mpid;
 
-	sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
-				  sizeof(arm64_cpu_boot_params));
+		sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
+					  sizeof(arm64_cpu_boot_params));
 
-	if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
-		printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
-		       cpu_num, cpu_mpid);
+		if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
+			printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
+			       cpu_num, cpu_mpid);
+			continue;
+		}
```

Comment:
In my opinion, I think trying to boot the next core if booting the previous core failed is not a good idea. The developer should be responsible for the results of booting the secondary core. If it failed, they should look at it rather than booting another available core from dts by Zephyr.

So if one core can not be booted, I would print an error message and return immediately.

How do you think @carlocaione @povergoing.

Reply:
agree. @quic-chadk what is your problem / use-case here?

Reply:
@carlocaione, @SgrrZhf   Sometimes a core or cluster of cores are intentionally not available.  Maybe it's because they are on an optional module that is not installed, maybe it's because of manufacturing yield.  For example, I believe it was the Sony PS3 Cell processor that was allowed to have 1 dead core but still be acceptable.  So  it is totally possible that you might have 4 cores but the system only needs 3 to function.  If one dies it is ok, just use the next available core.  Also sometimes devices are in hard to reach locations.  High up off the ground, out in remote locations.  In those situations you cannot just easily drive out and swap out the hardware.  The thinking is more commercial devices and not development on our desks.

Reply:
This behavior is very unusual and at the very minimum it must be clearly documented somewhere to not confuse people or even selectively enabled using a Kconfig option.

Reply:
> This behavior is very unusual and at the very minimum it must be clearly documented somewhere to not confuse people or even selectively enabled using a Kconfig option.

I appreciate the open mind, thank you.  I can modify the logic to make this behavior a Kconfig option.  The help section will document what it does and it will default disabled.  What would you like the name to be, ARM64_ENABLE_RESERVED_CORES?

Reply:
> `ARM64_FALLBACK_ON_RESERVED_CORES` maybe?

Sure, that name is fine.  I will make it so.  Thank you again.

Reply:
Changes are ready for review.  Thank you.